implement transient BtSG statistic

dbkeitel opened the pull request:
- implemented in `ComputeFstat` and `TransientGridSearch` classes
- based on [XLALComputeTransientBstat](https://lscsoft.docs.ligo.org/lalsuite/lalpulsar/group___transient_c_w__utils__h.html#gaf66cf2ffa017a6b1dffd266139b52fad) and Prix, Giampanis & Messenger 2011
- introduces a `ComputeFstat.get_transient_detstats()` method and makes `get_transient_detstats.get_transient_maxTwoFstat()` a legacy wrapper around it
- refs #260

Currently with `tCWFstatMapVersion="lal"`, the original lalpulsar function is called for BtSG as well. With `tCWFstatMapVersion="pycuda"`, the python version of BtSG is called after the pycuda computation of the FstatMap. A naive python port of that function (with double for loop) was very slow (BtSG took up to factor 10 longer than FstatMap), but a simple change to `np.sum()` was enough to make the extra cost negligible, and so it seems porting BtSG to cuda is not necessary.

TODO:

- [x] add to `TransientMCMCSearch`
- [x] add meaningful quantitative tests
- [x] fix pipeline
- [x] squash

dbkeitel commented:
cc @itsluanayall @ReinhardPrix

dbkeitel commented:
Did some necessary cleanup and extended tests. As per the resolved in-line thread above, using `scipy.special.logsumexp` seems counterproductive, and moving the marginalisation into cuda does not seem to be required either. So I think this is ready for review - let me assign @Rodrigo-Tenorio but also add @itsluanayall because there are enough chances for me having forgotten a factor 2 or something somewhere that 3 pairs of eyes will be ideal.

Rodrigo-Tenorio commented on pyfstat/tcw_fstat_map_funcs.py:
```diff
@@ -122,6 +124,33 @@ def get_maxF_idx(self):
+        # can be handled separately, eg by computing log(B) = Fmax + log(sum e^(Fmn-Fmax)) for the Bayes-factor.
+        sum_eB = np.sum(np.exp(-self.maxF - self.F_mn))
+        # combine this to final log(Bstat) result with proper normalization (assuming rhohMax=1):
+        logBhat = self.maxF + np.log(sum_eB)  # unnormalized Bhat
```
So... No logsumexp on implementation but yes to test against?

dbkeitel replied:
I've deliberately added a commit that introduces an extra function for it, and an immediate revert, so that we have this version in the history and can easily go back to testing against it if needed.

dbkeitel commented:
The codecov ` 85.48% of diff hit (target 70.78%)` doesn't look too great, but the missing lines are all exceptions, the `get_transient_maxTwoFstat` method I only kept around as a backwards compatibility helper, and the pycuda code that we currently can't test anyway.

I guess I could learn the syntax to tell codecov to ignore something, but for now it's above target already and hence happy.

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -661,16 +669,20 @@ def init_computefstatistic(self):
                 raise ValueError("Can't use BSGL with single detectors data")
             else:
                 logging.info("Initialising BSGL")
+            if getattr(self, "BtSG", False):
```
```suggestion
            if getattr(self, "BtSG", False): # pragma: no cover
```

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -1138,37 +1165,48 @@ def get_fullycoherent_log10BSGL(self):
             at the input parameter values.
-            Also stored as `self.maxTwoF`.
         """
+        if BtSG and BSGL:
```
```suggestion
        if BtSG and BSGL: # pragma: no cover
```

Rodrigo-Tenorio commented on pyfstat/grid_based_searches.py:
```diff
@@ -943,8 +946,12 @@ def __init__(
         for k in self.search_keys:
             setattr(self, k, np.atleast_1d(getattr(self, k + "s")))
-        if self.BSGL:
+        if self.BSGL and self.BtSG:
```
```suggestion
        if self.BSGL and self.BtSG: # pragma: no cover
```

Rodrigo-Tenorio commented on pyfstat/tcw_fstat_map_funcs.py:
```diff
@@ -87,10 +97,22 @@ def __init__(self, N_t0Range=None, N_tauRange=None, transientFstatMap_t=None):
+            If true, also compute the logBtSG transient Bayes factor statistic.
         """
-        if transientFstatMap_t:
+        if transientFstatMap_t and from_file:
```
```suggestion
        if transientFstatMap_t and from_file: # pragma: no cover
```

Rodrigo-Tenorio commented on pyfstat/tcw_fstat_map_funcs.py:
```diff
@@ -612,6 +716,11 @@ def pycuda_compute_transient_fstat_map(multiFstatAtoms, windowRange):
         )
     )
 
+    if BtSG:
```
```suggestion
    if BtSG: # pragma: no cover
```

dbkeitel replied:
Rejecting this one, this whole function _should_ be covered, including this case.

Rodrigo-Tenorio reviewed (changes requested):
I'll let you decide whether you want to add these or not :wink:

Rodrigo-Tenorio reviewed (approved):
Examples and __init__ look fine to me.

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -661,16 +669,20 @@ def init_computefstatistic(self):
                 raise ValueError("Can't use BSGL with single detectors data")
             else:
                 logging.info("Initialising BSGL")
+            if getattr(self, "BtSG", False):
```
Can this be moved earlier during initialization?

dbkeitel replied:
probably, will include in a cleanup commit

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -993,10 +1008,22 @@ def get_fullycoherent_detstat(
+
+        * maxTwoF (transient)
+
+        * logBtSG (transient)
```
```suggestion
        * log10BtSG (transient)
```
If not, why? Is `log10BSGL` historical and `logBtSG` ye new trend? In such case, `log` ---> `ln`.

dbkeitel replied:
Historical divergence in lalsuite conventions for the two statistics. I thought about normalising to log10, but then one always has to convert when wanting to compare to lalsuite outputs. Using the less ambiguous `ln` is a very good suggestion, though.

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -1033,11 +1059,12 @@ def get_fullycoherent_detstat(
+        return self.get_transient_detstats(
+            tstart=tstart,
+            tend=tend,
+            BSGL=getattr(self, "BSGL", False),
+            BtSG=getattr(self, "BtSG", False),
```
Can these be `self.BSGL=self.BtSG=False` by default at initialization time?

Rodrigo-Tenorio replied:
Actually, why give it as inputs if already present as attributes?

dbkeitel replied:
We're in a bit of a messy situation of sometimes allowing such methods to be called independently of what the setup was (see e.g. the test for this one), sometimes not. So giving the lower-level function these input options, despite its caller in the standard graph using the values from `self`, allows for such more flexible out-of-graph calls. It's probably not great practice, but kinda convenient...

dbkeitel replied:
Got rid of these as discussed.

Rodrigo-Tenorio commented on pyfstat/grid_based_searches.py:
```diff
@@ -50,7 +50,9 @@ class GridSearch(BaseSearchClass):
-        "log10BSGL": r"$\log_{10}\mathcal{B}_{\mathrm{SGL}}$",
+        "maxTwoF": r"$\max\widetilde{2\mathcal{F}}$",
+        "log10BSGL": r"$\log_{10}\mathcal{B}_{\mathrm{S/GL}}$",
+        "logBtSG": r"$\log\mathcal{B}_{\mathrm{tS/G}}$",
```
Same question re `log`.

Rodrigo-Tenorio commented on pyfstat/tcw_fstat_map_funcs.py:
```diff
@@ -122,11 +184,36 @@ def get_maxF_idx(self):
         return np.unravel_index(np.argmax(self.F_mn), self.F_mn.shape)
 
+    def get_logBtSG(self):
+        """Compute (natural log of the) transient-CW Bayes-factor B_tSG = P(x|HyptS)/P(x|HypG).
```
Ahà! So its natural log @dbkeitel : Why not `ln`?

Rodrigo-Tenorio replied:
Also... `HyptS`-->`Hyp[tS]`? `H_{tS}`?

dbkeitel replied:
Copied like this from lalpulsar comments. Version with underscore (and mathrm) would be the proper texy one.

Rodrigo-Tenorio commented on pyfstat/tcw_fstat_map_funcs.py:
```diff
@@ -122,11 +184,36 @@ def get_maxF_idx(self):
+        using given type and parameters of transient window range.
+
+        This is a python port of the `lalpulsar.ComputeTransientBstat` implementation,
+        replacing for loops by numpy operations.
```
```suggestion
        replacing `for loops by numpy operations.
```

dbkeitel commented:
Cleaned up and simplified the API after discussions with @Rodrigo-Tenorio . Now if people want to switch between computing lnBtSG and log10BSGL, they need to initialise a new ComputeFstat class. See `test_transient_detstats()` for a worked example.

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -699,6 +718,9 @@ def init_computefstatistic(self):
         if self.transientWindowType:
             logging.info("Initialising transient parameters")
+            self.maxTwoF = 0
+            if getattr(self, "BtSG", False):
```
```suggestion
            if self.BtSG:
```
I don't recall if there was a good reason for this specific one or we simply forgot to change it.
Attribute `BtSG` is always initialized, so no need for getattr.

dbkeitel replied:
Put it back in specifically because when called from semicoherent classes, it was not initialised.

Rodrigo-Tenorio commented on pyfstat/core.py:
```diff
@@ -1180,23 +1222,63 @@ def get_transient_maxTwoFstat(
         if np.isnan(self.maxTwoF):
             self.maxTwoF = 0
+
+        if getattr(self, "BtSG", False):
```
```suggestion
        if self.BtSG:
```

dbkeitel replied:
See above, needed to make semicoherent classes to run without explicitly setting it to False in those too.

Rodrigo-Tenorio reviewed (changes requested):
Overall consistent with what we discussed
